src/plugins/dashboard/public/attribute_service/attribute_service.tsx

```diff
@@ -0,0 +1,137 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import {
+  EmbeddableInput,
+  SavedObjectEmbeddableInput,
+  isSavedObjectEmbeddableInput,
+  IEmbeddable,
+} from '../embeddable_plugin';
+import { SavedObjectsClientContract, SimpleSavedObject, I18nStart } from '../../../../core/public';
+import {
+  SavedObjectSaveModal,
+  showSaveModal,
+  OnSaveProps,
+  SaveResult,
+} from '../../../saved_objects/public';
+
+/**
+ * The attribute service is a shared, generic service that embeddables can use to provide the functionality
+ * required to fulfill the requirements of the ReferenceOrValueEmbeddable interface. The attribute_service
+ * can also be used as a higher level wrapper to transform an embeddable input shape that references a saved object
+ * into an embeddable input shape that contains that saved object's attributes by value.
+ */
+export class AttributeService<
+  SavedObjectAttributes,
+  ValType extends EmbeddableInput & { attributes: SavedObjectAttributes },
+  RefType extends SavedObjectEmbeddableInput
+> {
+  constructor(
+    private type: string,
+    private savedObjectsClient: SavedObjectsClientContract,
+    private i18nContext: I18nStart['Context']
+  ) {}
+
+  public async unwrapAttributes(input: RefType | ValType): Promise<SavedObjectAttributes> {
+    if (this.inputIsRefType(input)) {
+      const savedObject: SimpleSavedObject<SavedObjectAttributes> = await this.savedObjectsClient.get<
+        SavedObjectAttributes
+      >(this.type, input.savedObjectId);
+      return savedObject.attributes;
+    }
+    return input.attributes;
+  }
+
+  public async wrapAttributes(
+    newAttributes: SavedObjectAttributes,
+    useRefType: boolean,
+    embeddable?: IEmbeddable
+  ): Promise<Omit<ValType | RefType, 'id'>> {
+    const savedObjectId =
+      embeddable && isSavedObjectEmbeddableInput(embeddable.getInput())
+        ? (embeddable.getInput() as SavedObjectEmbeddableInput).savedObjectId
+        : undefined;
+
+    if (useRefType) {
+      if (savedObjectId) {
+        await this.savedObjectsClient.update(this.type, savedObjectId, newAttributes);
+        return { savedObjectId } as RefType;
+      } else {
+        const savedItem = await this.savedObjectsClient.create(this.type, newAttributes);
```

Comment:
any error scenarios here to handle?

Reply:
Good call! I added a generic catch with toast similar to the ones used in dashboard and lens etc.